### PR TITLE
BGP LS IPv6 parsing support

### PR DESCRIPTION
**Pull request by Tinusf**
IPv6 pasing now works for "ip-reachability-tlv" and "ip-reach-prefix".

---
This change is [Reviewable](https://reviewable.io/reviews/exa-networks/exabgp/837)

### CODE REVIEW

**Comment by ruissalo** [comment]
Hi @Tinusf 
Thanks for sending this! :) 
How are you testing? Do you mind pasting some decoding examples?
I'll take a closer look at the code.
cheers

**Comment by Tinusf** [comment]
For testing I am running one instance of Exabgp on a router that is actually in use at my job (Uninett) and another instance that is on my server. So I am testing with real data in a real environment. Example before parsing: 
```json
{
    "code":4,
    "parsed":false,
    "raw":"043C0200000000000000000100001A02000004000000E0020100048027002B02030006128039000034010900117E20010700000025190000000000000000"
}
```
Example after parsing:
```json
{
    "ls-nlri-type":4,
    "l3-routing-topology":0,
    "protocol-id":2,
    "node-descriptors":{
        "autonomous-system":224,
        "bgp-ls-identifier":"2150039595",
        "router-id":"128039000041"
    },
    "ip-reachability-tlv":"2001:700:0:2533::",
    "ip-reach-prefix":"2001:700:0:2533::/126",
    "nexthop":"128.39.0.43"
}
```

File: qa/tests/bgpls/tlvs.py [hunk 1]
```diff
@@ -9,10 +9,15 @@
 
 class TestTlvs(unittest.TestCase):
 
-    def test_ip_reach(self,):
+    def test_ip_reach_ipv4(self,):
         data = b'\n\n\x00'
-        tlv = IpReach.unpack(data)
-        self.assertEqual(tlv.json(), '"ip-reachability-tlv": "10.0.0.0"')
+        tlv = IpReach.unpack(data, 3)
+        self.assertEqual(tlv.json(), '"ip-reachability-tlv": "10.0.0.0", "ip-reach-prefix": "10.0.0.0/10"')
```
Comment (Tinusf) [inline qa/tests/bgpls/tlvs.py]: I broke this test on my last pull request.

**Comment by thomas-mangin** [comment]
should we add the address family for the decoded route in the JSON to help people parsing it ?

**Comment by thomas-mangin** [comment]
For the EOR message it is called `"afi"`, so I would use the name naming convention, also I would keep the name in lower case - using the AFI class (`AFI(value).name()`) for the representation (again for consistency).

**Comment by thomas-mangin** [comment]
Also import ipaddress looks like a new requirement as this is not available on python 2.x, could you use some of exabgp code or write some code which does not add a deps please ?

**Comment by ruissalo** [comment]
Using the same naming convention as @thomas-mangin suggested is a good idea. I wouldn't use
{"address-family":"IPv6"}/ {"address-family":"IPv4"} though, since those aren't BGP-LS NLRIs. At the moment the actual NLRI is enconded as "ls-nlri-type":4, "ls-nlri-type":3, etc. but the meaning is indeed cryptic. 
What do you think of using something like ```bgpls-prefix-v4/v6```,  ```bgpls-link```,  ```bgpls-node```, or similar? You may have to update the NAME attribute.

**Comment by thomas-mangin** [comment]
@ruissalo you came up with the current naming convention for ls. Happy to follow what you believe is best.

**Comment by Tinusf** [comment]
Where is the code that does the IPv6 compression? And have you guys decided on a name?

**Comment by ruissalo** [comment]
Hi @Tinusf ! Let us know if you need any help to close this PR :)
Do you want to add the patch to update the nlri name?
You use bgpls-prefix-v4/v6, bgpls-link and bgpls-node

**Comment by thomas-mangin** [comment]
@ruissalo I think @Tinusf is waiting for confirmation on how he should add the afi/safi information - I gave my preference - is it ok with you ?

**Comment by Tinusf** [comment]
Yeah, sorry I've been busy lately with other stuff. But I was also wondering where the exabgp code for compressing ipv6 addresses are so that I don't have to use import ipaddress

**Comment by thomas-mangin** [comment]
Do not worry and use it: `from exabgp.vendoring import ipaddress`
It is already there ..

**Comment by ruissalo** [comment]
@thomas-mangin sure, SGTM.

**Comment by Tinusf** [comment]
Something like this?
examples:

```json
{
    "ls-nlri-type": 1,
    "l3-routing-topology": 0,
    "afi": "bgpls-node",
    "nexthop": "128.39.0.43",
    "node-descriptors": {
        "autonomous-system": 224,
        "bgp-ls-identifier": "2150039595",
        "router-id": "128039000067"
    },
    "protocol-id": 2
}
{
    "local-node-descriptors": {
        "autonomous-system": 224,
        "bgp-ls-identifier": "2150039595",
        "router-id": "128039000067"
    },
    "ls-nlri-type": 2,
    "afi": "bgpls-link",
    "interface-address": {
        "interface-address": "128.39.254.131"
    },
    "remote-node-descriptors": {
        "autonomous-system": 224,
        "bgp-ls-identifier": "2150039595",
        "router-id": "128039000116"
    },
    "l3-routing-topology": 0,
    "protocol-id": 2,
    "neighbor-address": {
        "neighbor-address": "128.39.254.130"
    }
}

{
    "ip-reach-prefix": "128.39.254.48/31",
    "ls-nlri-type": 3,
    "nexthop": "128.39.0.43",
    "afi": "bgpls-prefix-v4",
    "l3-routing-topology": 0,
    "node-descriptors": {
        "autonomous-system": 224,
        "bgp-ls-identifier": "2150039595",
        "router-id": "128039000037"
    },
    "ip-reachability-tlv": "128.39.254.48",
    "protocol-id": 2
}
{
    "ip-reach-prefix": "2001:700:0:2049::/127",
    "ls-nlri-type": 4,
    "nexthop": "128.39.0.43",
    "afi": "bgpls-prefix-v6",
    "l3-routing-topology": 0,
    "node-descriptors": {
        "autonomous-system": 224,
        "bgp-ls-identifier": "2150039595",
        "router-id": "128039000037"
    },
    "ip-reachability-tlv": "2001:700:0:2049::",
    "protocol-id": 2
}
```

**Comment by thomas-mangin** [comment]
What you presenting as "afi" are neither AFI or SAFI and therefore can not be called that way.
you could have something like `"family": { "afi": "ipv6", "safi": "unicast" }`  or `"afi": "ipv4", "safi": "unicast"` 

The valid AFI / SAFI are on 
 https://www.iana.org/assignments/address-family-numbers/address-family-numbers.xhtml
 https://www.iana.org/assignments/safi-namespace/safi-namespace.xhtml

I am not sure what `bgpls-node` or `bgpls-node` is as I have not read the RFC yet ..
Do I make sense, or do I need to get up to date with RFC 7752 more ?

**Comment by ruissalo** [comment]
Hi @Tinusf , @thomas-mangin is right, but you can blame me for the confusion.

you could do something like:

`"family": { "afi": "bgp-ls", "safi": "bgp-ls" }` to be RFC compliant. The confusion comes from the fact that bgp-ls is a bit special in the sense that the same nlri has multiple `types`, which I don't think happens with other NLRIs. In the code, link, prefix and node bgp-ls nlris are registered as different NLRIs, but you could use the same name and short name for all three, leaving the CODE unique so they can be registered correctly. 

Where I think `bgpls-prefix-v6`, `bgpls-node` etc could be really useful is  to replace
` "ls-nlri-type": 3`,  `"ls-nlri-type": 2`, etc, which is kind of cryptic and requires knowledge of the rfc to know what it means. so you could change that to `"ls-nlri-type": "bgpls-prefix-v4"` etc.

thanks!

**Comment by thomas-mangin** [comment]
I was just wanting the safi in the ip-reach-prefix to be somewhere in the json so the parser did not have to "guess" from the data format what it is but could just get it from the json and know what to do

**Comment by Tinusf** [comment]
I changed from ```"ls-nlri-type": 3``` to ```"ls-nlri-type": "bgpls-prefix-v4"``` now. I did not add any afi or safi because I don't think I understand it completely. Maybe you guys could add it, I think that would be a lot easier.

**Comment by thomas-mangin** [comment]
@ruissalo the patch still applies .. should I commit ?

**Comment by ruissalo** [comment]
@thomas-mangin the patch should be OK, I never had the chance to test it with a router but given the unit test, it should be ok.

**Comment by thomas-mangin** [comment]
Just took 1 and 1/2 year to merge ... 🤦‍♂
Thank you @Tinusf and @ruissalo

**Comment by Tinusf** [comment]
> 
> 
> Just took 1 and 1/2 year to merge ... 🤦‍♂
> Thank you @Tinusf and @ruissalo

Haha, yea took a little while.

**Pull request merged**